apps/oxlint/src/output_formatter/gitlab.rs

```diff
@@ -0,0 +1,129 @@
+use oxc_diagnostics::{
+    Error, Severity,
+    reporter::{DiagnosticReporter, DiagnosticResult, Info},
+};
+
+use std::hash::{DefaultHasher, Hash, Hasher};
+
+use crate::output_formatter::InternalFormatter;
+
+#[derive(Debug, Default)]
+pub struct GitlabOutputFormatter;
+
+#[derive(Debug, serde::Serialize)]
+struct GitlabErrorLocationLinesJson {
+    begin: usize,
+    end: usize,
+}
+
+#[derive(Debug, serde::Serialize)]
+struct GitlabErrorLocationJson {
+    path: String,
+    lines: GitlabErrorLocationLinesJson,
+}
+
+#[derive(Debug, serde::Serialize)]
+struct GitlabErrorJson {
+    description: String,
+    check_name: String,
+    fingerprint: String,
+    severity: String,
+    location: GitlabErrorLocationJson,
+}
+
+impl InternalFormatter for GitlabOutputFormatter {
+    fn get_diagnostic_reporter(&self) -> Box<dyn DiagnosticReporter> {
+        Box::new(GitlabReporter::default())
+    }
+}
+
+/// Renders reports as a Gitlab Code Quality Report
+///
+/// <https://docs.gitlab.com/ci/testing/code_quality/#code-quality-report-format>
+///
+/// Note that, due to syntactic restrictions of JSON arrays, this reporter waits until all
+/// diagnostics have been reported before writing them to the output stream.
```

Comment:
note there're are optimizations we can do here, but i think this is fine for now

e.g, at start write `[`, then for each diagnostic as it arrives (skip comma for the first one) `format!(",{}")`, then on finish, write the final `]`

Reply:
Thanks for the tip! This is all still new to me, so I'm sure there's plenty of room for improvement :)